Fix naming of unittest classes and file names

AndrejMitrovic opened the pull request:
I often find myself looking at the wrong module when trying to find the Network class. I think this naming convention is easier for humans to distinguish between test modules and core modules, even if it's a little bit repetitive.

AndrejMitrovic committed:
Remove Network unittest subclass

The waitUntilConnected() method is useful
to have for most TestNetwork subclasses,
it can belong in the TestNetwork class.

Geod24 commented:
Hum, okay, not fan of the repetition, but there seem to be a consensus among the team on that.
Note that I expect that subclassing `TestNetwork` will be a quite common operation. The reason is that `TestNetwork` is the place where the node behavior is defined (as in, do we provide a byzantine node, a regular node, etc...)

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -83,6 +83,48 @@ public class TestNetwork : Network
+    /// Wait a certain time until the nodes have reached discovery
+    /// If after 10 query attempts they still all haven't discovered => assert
+    /// Params:
+    ///   count = Expected number of nodes
```
Too many lines for `///`

AndrejMitrovic replied:
I guess we are going to use `/** */` everywhere for multiline comments? I'm 50% sure that using multiline `///` comments works too. Anyway I will fix it.

Geod24 commented on source/agora/test/TestNetwork.d:
```diff
@@ -0,0 +1,35 @@
+
+import std.algorithm.iteration;
+import std.exception;
+import std.format;
```
All those import don't seem necessary ?

AndrejMitrovic commented:
> Hum, okay, not fan of the repetition, but there seem to be a consensus among the team on that.
Note that I expect that subclassing TestNetwork will be a quite common operation.

There is a small usability issue with this. When you do inherit from `TestNetwork`, you must ensure to override `createNewNode`, otherwise the node that's instantiated will still be templated on the base class. For example in here:

```D
/// Initialize a new node
protected TestAPI createNewNode (Address address, Config conf)
{
    auto api = RemoteAPI!TestAPI.spawn!(TestNode!TestNetwork)(conf);
    std.concurrency.register(address, api.tid());
    return api;
}
```

In the derived class you would have to override createNewNode to do `spawn!(TestNode!DerivedTestNetwork)`. I ran into it a few times already.

Geod24 commented:
True, that's not ideal. Ideally I'd like to get rid of the template arg and just instantiate it from outside the class though.

AndrejMitrovic commented:
Aha so something like `getNetwork()` inside of the Node?

Geod24 commented:
I have to look at alternatives, but more like passing `Network` to the ctor, and having `LocalRest` support an instantiating function instead of `new Type`

AndrejMitrovic commented:
Hmm there is one issue, `waitUntilConnected` will assert in a place other than the function call. Let me rework this a little bit.

AndrejMitrovic committed:
Move starting the nodes into TestNetwork

The makeTestNetwork() helper function should
only create the network, it shouldn't start
any nodes, that is up to the network itself.

AndrejMitrovic committed:
Get rid of count param in waitUntilConnected()

This can be derived from the length of the apis array.

AndrejMitrovic committed:
Rename waitUntilConnected() to getDiscoveredNodes()

Also make it return the connected nodes public keys,
rather than asserting internally. This ensures
assertion errors happen at the point of invocation.

AndrejMitrovic commented:
I've added 3 more commits. Maybe you want me to split this up into two PRs, or is that overkill?

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -104,12 +104,9 @@ public class TestNetwork : Network
     ***************************************************************************/
 
-    public void waitUntilConnected (size_t count)
+    public void waitUntilConnected ()
```
Well the point of doing this is that, if for some reason the `apis` array is empty, this just passes.

AndrejMitrovic replied:
Why should it pass?

Geod24 replied:
You'll enter the `foreach` for `attempts`. Then the `foreach` on `this.apis` will not be executed because `this.apis.length == 0`. Then L132 will pass (0 == 0) and it returns.

AndrejMitrovic replied:
I mean the `assert(network.getDiscoveredNodes().length == NodeCount);` will fail. And then you get the assert message where it's used.

Geod24 replied:
But then you rely on the caller checking that you actually connected to all the nodes.

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -139,9 +141,7 @@ public class TestNetwork : Network
 
-        assert(fully_discovered.length == count,
-               format("Got %s/%s discovered nodes. Missing nodes: %s",
-                   fully_discovered.length, this.apis.length, this.todo_addresses));
```
Honestly I don't see much of an improvement here... We do have stack traces so the assert being here shouldn't be a problem?

AndrejMitrovic replied:
Aren't stack traces still broken?

AndrejMitrovic replied:
```
core.exception.AssertError@source/agora/test/Base.d(145): Got 4/5 discovered nodes. Missing nodes: ["GD6GUK2NDHNTAS5TAKCIE6BFSLG5WELOE2GLRMZTORCY4FPWZOY5EPAZ":true, "GCI6VCQJG53N3HWQVENP6WED56YREJTDRQOW2VC2UBEADCST42GDE3QB":true, "GBDESH4FGFCEMTDADJMCFKPCHHZSFKZ4U5BYLX6NIB6ED3RDVZ23UE2Z":true, "GBCJQKB4QRTPUX4Y2H6PAKKRKIKH7RV4K5PK5CGTUI2EDH2SD3NAQY7Z":true]
----------------
??:? _d_assert_msg [0x1026c86c6]
/usr/local/opt/dmd/include/dlang/dmd/object.d:2911 void agora.test.Base.TestNetwork.waitUntilConnected(ulong) [0x1023c8051]
/usr/local/opt/dmd/include/dlang/dmd/object.d:2911 void agora.test.TestNetwork.__unittest_L19_C1() [0x1023ca805]
source/agora/app.d:36 void agora.test.TestNetwork.__modtest() [0x10234e5e8]
??:? int core.runtime.runModuleUnitTests().__foreachbody2(object.ModuleInfo*) [0x1026c93fb]
??:? int object.ModuleInfo.opApply(scope int delegate(object.ModuleInfo*)).__lambda2(immutable(object.ModuleInfo*)) [0x1026bd9ce]
```

Yeah they're still broken.

Geod24 replied:
They are, but moving this responsibility to the caller does not appeal to me too much... But okay, let's give it a try.

AndrejMitrovic commented:
@Geod24 I think Michael's only issue was with the subclass `Network => TestNetwork => Network`, so I could revert the commit which renames the files. That alright?

AndrejMitrovic commented:
Reverted that commit.

Geod24 commented on source/agora/test/Network.d:
```diff
@@ -13,76 +13,11 @@
 
 module agora.test.Network;
 
-version (unittest):
```
That should stay though

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -84,6 +84,20 @@ public class TestNetwork : Network
+        Start each of the nodes
+
+        Params:
+            count = Expected number of nodes
```
What ? :D

AndrejMitrovic replied:
This was moved from one place to another. And also, I remove it in a separate commit!

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -84,6 +84,20 @@ public class TestNetwork : Network
+
+    ***************************************************************************/
+
+    public void startNodes ()
```
Maybe just `start`?

AndrejMitrovic commented:
Applied fixes.

Geod24 reviewed (approved):
Heh let's try
